api/solarperformanceinsight_api/compute.py

```diff
@@ -30,22 +35,326 @@ def run_job(job_id: UUID, user: str):
+            details = str(err.args[0])
+        except IndexError:
+            details = f"Raised {type(err)}"
+        msg = json.dumps({"error": {"details": details}})
         with si.start_transaction() as st:
-            st.add_job_result(
-                job_id,
-                "/",
-                "error message",
-                "application/json",
-                json.dumps({"error": {"details": err.args[0]}}),
-            )
+            st.add_job_result(job_id, "/", "error message", "application/json", msg)
             st.set_job_error(job_id)
 
 
 def lookup_job_compute_function(
     job: models.StoredJob,
 ) -> Callable[[models.StoredJob, storage.StorageInterface], None]:
+    if isinstance(job.definition.parameters.job_type, models.CalculatePerformanceJob):
+        return run_performance_job
     return dummy_func
 
 
-def dummy_func(job, storage):
+def dummy_func(job, storage):  # pragma: no cover
     raise NotImplementedError("Job computation not implemented")
+
+
+def _get_data(
+    job_id: UUID, data_id: UUID, si: storage.StorageInterface, shift: dt.timedelta
+) -> pd.DataFrame:
+    """Get the data from the database and shift forward by `shift`"""
+    with si.start_transaction() as st:
+        meta, data = st.get_job_data(job_id, data_id)
+    if meta.definition.data_format != "application/vnd.apache.arrow.file":
+        raise TypeError(
+            f"Data for /jobs/{job_id}/data/{data_id} not in Apache Arrow format"
+        )
+    return utils.read_arrow(data).set_index("time").shift(freq=shift)  # type: ignore
+
+
+def generate_job_weather_data(
+    job: models.StoredJob, si: storage.StorageInterface, shift: dt.timedelta
+) -> Generator[List[pd.DataFrame], None, None]:
+    """Generator to fetch job data at the inverter level to run a ModelChain"""
+    data_id_by_schema_path = {
+        do.definition.schema_path: do.object_id
+        for do in job.data_objects
+        if do.definition.type
+        in (
+            models.JobDataTypeEnum.original_weather,
+            models.JobDataTypeEnum.actual_weather,
+        )
+    }
+    job_id = job.object_id
+    num_inverters = len(job.definition.system_definition.inverters)
+    if (
+        job.definition.parameters.weather_granularity
+        == models.WeatherGranularityEnum.system
+    ):
+        data_id = data_id_by_schema_path["/"]
+        df = _get_data(job_id, data_id, si, shift)
+        for i in range(num_inverters):
+            num_arrays = len(job.definition.system_definition.inverters[i].arrays)
+            yield [df.copy()] * num_arrays
+    elif (
+        job.definition.parameters.weather_granularity
+        == models.WeatherGranularityEnum.inverter
+    ):
+        for i in range(num_inverters):
+            num_arrays = len(job.definition.system_definition.inverters[i].arrays)
+            data_id = data_id_by_schema_path[f"/inverters/{i}"]
+            df = _get_data(job_id, data_id, si, shift)
+            yield [df] * num_arrays
```

Comment:
maybe these need to be copies of df depending on if pvlib does any modifying of the frames

Reply:
block above reuses the same dataframe on each iteration vs getting data for each iteration here. We won't have to worry about copying here for our current use since the later `.shift` will produce a copy